File: backend/app/models/intelligence.py
from datetime import UTC, datetime, timedelta
from typing import Literal
from zoneinfo import ZoneInfo, ZoneInfoNotFoundError

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator


class TrendGenerateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    window_end: datetime
    window_hours: int = Field(default=24, ge=6, le=168)

    @field_validator("window_end")
    @classmethod
    def utc_window_end(cls, value: datetime) -> datetime:
        return value.replace(tzinfo=UTC) if value.tzinfo is None else value.astimezone(UTC)


class TrendSignal(BaseModel):
    id: str
    topic: str
    window_start: str
    window_end: str
    score: float
    volume_score: float
    velocity_score: float
    cross_platform_score: float
    engagement_score: float
    platforms: list[str]
    content_ids: list[str]
    explanation: str
    evidence: dict[str, object]
    status: Literal["detected", "insufficient_data"]
    formula_version: str
    created_at: str


class BriefGenerateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    window_start: datetime
    window_end: datetime
    timezone: str = Field(min_length=1, max_length=100)
    regenerate: bool = False

    @field_validator("window_start", "window_end")
    @classmethod
    def utc_timestamp(cls, value: datetime) -> datetime:
        return value.replace(tzinfo=UTC) if value.tzinfo is None else value.astimezone(UTC)

    @field_validator("timezone")
    @classmethod
    def valid_timezone(cls, value: str) -> str:
        try:
            ZoneInfo(value)
        except ZoneInfoNotFoundError as error:
            raise ValueError("timezone must be a valid IANA timezone") from error
        return value

    @model_validator(mode="after")
    def ordered_window(self) -> "BriefGenerateRequest":
        if self.window_end <= self.window_start:
            raise ValueError("window_end must be after window_start")
        if self.window_end - self.window_start > timedelta(days=31):
            raise ValueError("brief window must not exceed 31 days")
        return self


class BriefItem(BaseModel):
    id: str
    section: str
    conclusion_type: Literal[
        "fact",
        "calculation",
        "rule",
        "insufficient_data",
        "unknown",
    ]
    title: str
    body: str
    position: int
    evidence: dict[str, object]
    content_ids: list[str]
    trend_ids: list[str]


class Brief(BaseModel):
    id: str
    window_start: str
    window_end: str
    timezone: str
    version: int
    generator: Literal["deterministic", "ai_enhanced"]
    ai_provider: str
    status: str
    created_at: str
    evidence_count: int
    items: list[BriefItem]


class BriefScheduleWrite(BaseModel):
    model_config = ConfigDict(extra="forbid")

    enabled: bool
    timezone: str = Field(min_length=1, max_length=100)
    time_of_day: str = Field(pattern=r"^(?:[01]\d|2[0-3]):[0-5]\d$")

    @field_validator("timezone")
    @classmethod
    def schedule_timezone(cls, value: str) -> str:
        try:
            ZoneInfo(value)
        except ZoneInfoNotFoundError as error:
            raise ValueError("timezone must be a valid IANA timezone") from error
        return value


class BriefSchedule(BriefScheduleWrite):
    id: str
    last_run_at: str | None
    next_run_at: str | None
    consecutive_failures: int
    last_error: str | None
    created_at: str
    updated_at: str

File: frontend/src/api/intelligence.ts
import { z } from "zod";

import { requestJson } from "./client";

export const trendSchema = z.object({
  id: z.string(),
  topic: z.string(),
  window_start: z.string(),
  window_end: z.string(),
  score: z.number(),
  volume_score: z.number(),
  velocity_score: z.number(),
  cross_platform_score: z.number(),
  engagement_score: z.number(),
  platforms: z.array(z.string()),
  content_ids: z.array(z.string()),
  explanation: z.string(),
  evidence: z.record(z.string(), z.unknown()),
  status: z.enum(["detected", "insufficient_data"]),
  formula_version: z.string(),
  created_at: z.string(),
});

const briefItemSchema = z.object({
  id: z.string(),
  section: z.string(),
  conclusion_type: z.enum([
    "fact",
    "calculation",
    "rule",
    "insufficient_data",
    "unknown",
  ]),
  title: z.string(),
  body: z.string(),
  position: z.number().int(),
  evidence: z.record(z.string(), z.unknown()),
  content_ids: z.array(z.string()),
  trend_ids: z.array(z.string()),
});

export const briefSchema = z.object({
  id: z.string(),
  window_start: z.string(),
  window_end: z.string(),
  timezone: z.string(),
  version: z.number().int().positive(),
  generator: z.enum(["deterministic", "ai_enhanced"]),
  ai_provider: z.string(),
  status: z.string(),
  created_at: z.string(),
  evidence_count: z.number().int().nonnegative(),
  items: z.array(briefItemSchema),
});

const briefScheduleSchema = z.object({
  id: z.string(),
  enabled: z.boolean(),
  timezone: z.string(),
  time_of_day: z.string(),
  last_run_at: z.string().nullable(),
  next_run_at: z.string().nullable(),
  consecutive_failures: z.number().int().nonnegative(),
  last_error: z.string().nullable(),
  created_at: z.string(),
  updated_at: z.string(),
});

export type Trend = z.infer<typeof trendSchema>;
export type Brief = z.infer<typeof briefSchema>;

export function listTrends(signal?: AbortSignal): Promise<Trend[]> {
  return requestJson("/api/intelligence/trends", z.array(trendSchema), {
    signal,
  });
}

export function getLatestBrief(signal?: AbortSignal): Promise<Brief> {
  return requestJson("/api/intelligence/briefs/latest", briefSchema, {
    signal,
  });
}

export function generateTrends(): Promise<Trend[]> {
  return requestJson("/api/intelligence/trends/generate", z.array(trendSchema), {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      window_end: new Date().toISOString(),
      window_hours: 24,
    }),
  });
}

export function generateBrief(regenerate = false): Promise<Brief> {
  const end = new Date();
  const start = new Date(end.getTime() - 24 * 60 * 60 * 1000);
  return requestJson("/api/intelligence/briefs", briefSchema, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      window_start: start.toISOString(),
      window_end: end.toISOString(),
      timezone: Intl.DateTimeFormat().resolvedOptions().timeZone || "UTC",
      regenerate,
    }),
  });
}

export function getBriefSchedule(signal?: AbortSignal) {
  return requestJson(
    "/api/intelligence/briefs/schedule",
    briefScheduleSchema,
    { signal },
  );
}

export function setBriefSchedule(input: {
  enabled: boolean;
  timezone: string;
  time_of_day: string;
}) {
  return requestJson(
    "/api/intelligence/briefs/schedule",
    briefScheduleSchema,
    {
      method: "PUT",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(input),
    },
  );
}
